hail-is/hail
# [many] TLS

This PR teaches most of our cluster how to exclusively speak HTTPS instead of
HTTP. The exceptions are:
- from batch-driver to batch workers
- from batch workers to internal-gateway
- to ukbb-rg
- from router to notebook workers
- letsencrypt (oh the irony)

The major new build step is `create_certs` which creates a certificate, key, and
list of trusted "principals" for each "principal". "Principal" is a computer
security term referring to an authenticatable identity. In our system, the
services are each unique principals and every client (e.g. the test_batch CI
step) is also a principal. A principal's certificate is a unforgeable proof of
their identity. A principal's "key", in our system, is actually a public-private
(i.e. asymmetric) key pair which the client and server use to establish a
symmetric key for each new connection. A list of trusted principals is a list of
certificates. Every incoming connection must provide a certificate in the
trusted list or the server will drop the connection.

Every service depends on the `create_certs` step because their deployment's load
secrets created by `create_certs`.

The blog service is implemented by Ghost. Ghost only supports HTTP. As a result
we cannot make all network traffic in our cluster TLS-secured. However, we can
use an nginx sidecar on the blog pod which terminates TLS connections and sends
plaintext traffic on the loopback interface to Ghost. Thus, our goal is: no
plaintext traffic on non-loopback interfaces.

Readiness and liveness probes cannot use HTTP. Although k8s supports HTTPS, it
does not support so-called "mTLS" or "mutual TLS." This is fancy verbiage for
servers that require clients to send trusted certificates. We require
this. There is a lot of information in GitHub issues and the Istio web pages
about this, but at the end of the day, kubernetes does not support this. TCP
probes are the best we can do. There [was a
PR](https://github.com/kubernetes/kubernetes/pull/61231) to allow httpGet probes
to send the kubelet certificate, but it was closed because, apparently, the
[httpGet probes can be targeted at arbitrary IP
addresses](https://github.com/kubernetes/kubernetes/pull/61231#pullrequestreview-104364784) (what
the hell?), ergo Confused Deputy.

I also partly resolved the batch [confused deputy
problem](https://en.wikipedia.org/wiki/Confused_deputy_problem)! Batch will only
use its TLS identity when making callbacks for CI jobs. This makes CI quite
powerful, but we control it. All other batch users cannot use batch callbacks to
trick batch into issuing HTTP(S) requests to random services in our system
(because those services will reject a request from an untrusted principal).

Note that the internal-gateway still only accepts HTTP for incoming
requests. Outgoing requests to the router are HTTPS.

---

New concepts:
- a type of secret `ssl-config-{principal}`
- a global configuration file: `tls/config.yaml`

The new secret must have:
- `{principal}-key.pem`: the principal's private key (identity).
- `{principal}-cert.pem`: the principal's certificate (proof of identity).
- `{principal}-trust.pem`: a list of "trusted" certificates.
and additionally must have one of:
- `ssl-config.json`: a Hail configuration file in json format.
- `ssl-config-http.conf` and `ssl-config-proxy.conf`: NGINX configuration files
  that configure server-side TLS in the `http` block and client-side
  (i.e. proxy-side) TLS in a location block.
- `ssl-config.curlrc`: a curl configuration file.
A [PEM file](https://en.wikipedia.org/wiki/Privacy-Enhanced_Mail) is a
base64 encoding standard for certificates and keys. Our certificates are
standard TLS [X.509 certificates](https://en.wikipedia.org/wiki/X.509). Our
private keys are RSA 4096 keys.

The configuration file lists every principal in the system and the "ssl-mode" of
the system. The "ssl-mode" is inspired by MySQL's ssl-mode's and is one of (in
order from most to least secure): VERIFY_CA, REQUIRED, DISABLED.

|mode | incoming connections must use TLS | clients verify hostnames on server cert | servers only accept trusted clients |
|---|---|---|---|
|VERIFY_CA|yes|yes|yes|
|REQUIRED|yes|no|no|
|DISABLED|no|no|no|

`create_certs.py` converts this global configuration file into a secret for each
principal.

For NGINX principals, we generate the nginx conf in
`create_certs.py`. Unfortunately, I have no simple way to change the ports and
`ssl` status on nginx servers. For DISABLED, we send empty configuration
files. For REQUIRED, we load server certs and client certs, but we do not verify
(proxied) servers. I load the client certificates anyway so that I can smoke
test them before I require servers verify them. For VERIFY_CA, we load server
certs, load client certs, verify clients, and verify (proxied) servers.

For Hail principals, we only generate a json configuration
file containing the ssl mode and some named paths. The new `hailtop/ssl.py`
module defines the mapping from configuration to Python's
[`SSLContext`](https://docs.python.org/3.6/library/ssl.html#ssl.SSLContext).

There is also one "curl" principal: the admin-pod. REQUIRED and DISABLED are
mostly the same because required passes the `insecure` flag. As curl is
client-only, there is no notion of "incoming connection".

---

FAQ

Does this recreate certs on each deploy?

Yes.

How do services speak to each other while a deploy is happening? Newly deployed
services will only trust newly deployed clients?

`create_certs.py` includes the previous deploy's certificates in the trust
chain, so we can always accept clients from one deploy backwards. Old services
will not trust the new clients, but the `build.yaml` ensures things are deployed
in dependency order. Deploy would never work if a client could depend on
a not-yet-deployed server.

---

Things for you to verify:
- does every service load *its own unqiue* ssl-config secret?
- does every service use an SSL context for serving?
- does everyone who makes http requests to internal services use an SSL client
  session?
- do all TLS-secured nginx instances include their http.conf in the `http`
  section and the `proxy.conf` file in any proxying sections?
- Do the SSLContext's from `ssl.py` and the nginx configurations generated by
  `create_certs.py` obey the aforementioned matrix?

Post-merge actions:
- deploy gateway
- deploy internal-gateway
- deploy router-resolver

Anticipated outages:

- Before a service is redeployed it will be inaccessible from the outside
  because the router will try to speak to it on HTTPS. Services that speak to
  one another (ci<->batch, everyone<->auth) will lose connections while the
  deploy is happening. Deploy should move smoothly because CI will completely
  transmit the deploy batch to batch before batch goes dark.
- dev namespaces will be broken until the owner redeploys the router, etc.
